diff --git a/terraform/environments/bootstrap/member-bootstrap/iam.tf b/terraform/environments/bootstrap/member-bootstrap/iam.tf
index 7138de2cf..16218e1d9 100644
--- a/terraform/environments/bootstrap/member-bootstrap/iam.tf
+++ b/terraform/environments/bootstrap/member-bootstrap/iam.tf
@@ -49,7 +49,6 @@ data "aws_iam_policy_document" "member-access" {
       "codebuild:*",
       "codedeploy:*",
       "codepipeline:*",
-      "cognito:*",
       "cognito-idp:*",
       "datasync:*",
       "dbqms:*",
@@ -82,15 +81,17 @@ data "aws_iam_policy_document" "member-access" {
       "ec2:DeleteVpc",
       "ec2:CreateSubnet",
       "ec2:DeleteSubnet",
+      "ec2:AssociateRouteTable",
+      "ec2:CreateRouteTable",
+      "ec2:DeleteRouteTable",
       "ec2:CreateVpcEndpoint",
       "ec2:DeleteVpcEndpoints",
       "ec2:CreateVpcEndpointServiceConfiguration",
       "ec2:DeleteVpcEndpointServiceConfigurations",
-      "ec2:ModifyVpcEndpointServiceConfiguration",
-      "ec2:DescribeVpcEndpointServiceConfigurations",
-      "ec2:DescribeVpcEndpointServicePermissions",
-      "ec2:ModifyVpcEndpointServicePermissions",
+      "ec2:ModifyVpc*",
       "ec2:Describe*",
+      "ec2:*NetworkAcl*",
+      "ec2:*FlowLogs",
       "ec2:*SecurityGroup*",
       "ec2:*KeyPair*",
       "ec2:*Tags*",
