diff --git a/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/ProjectionColumnProcessor.java b/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/ProjectionColumnProcessor.java
index 3dde9b20..a27af237 100644
--- a/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/ProjectionColumnProcessor.java
+++ b/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/ProjectionColumnProcessor.java
@@ -30,6 +30,7 @@ import org.slf4j.LoggerFactory;
 
 import java.lang.reflect.InvocationTargetException;
 import java.util.ArrayList;
+import java.util.LinkedHashSet;
 import java.util.List;
 
 /**
@@ -98,7 +99,9 @@ public class ProjectionColumnProcessor {
 
         // 1 - Add referenced columns
         RecordData.FieldGetter[] fieldGetters = tableInfo.getPreTransformedFieldGetters();
-        for (String originalColumnName : projectionColumn.getOriginalColumnNames()) {
+        LinkedHashSet<String> originalColumnNames =
+                new LinkedHashSet<>(projectionColumn.getOriginalColumnNames());
+        for (String originalColumnName : originalColumnNames) {
             switch (originalColumnName) {
                 case TransformParser.DEFAULT_NAMESPACE_NAME:
                     params.add(tableInfo.getNamespace());
@@ -142,7 +145,8 @@ public class ProjectionColumnProcessor {
         List<Class<?>> paramTypes = new ArrayList<>();
         List<Column> columns = tableInfo.getPreTransformedSchema().getColumns();
         String scriptExpression = projectionColumn.getScriptExpression();
-        List<String> originalColumnNames = projectionColumn.getOriginalColumnNames();
+        LinkedHashSet<String> originalColumnNames =
+                new LinkedHashSet<>(projectionColumn.getOriginalColumnNames());
         for (String originalColumnName : originalColumnNames) {
             for (Column column : columns) {
                 if (column.getName().equals(originalColumnName)) {
diff --git a/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/TransformFilterProcessor.java b/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/TransformFilterProcessor.java
index 84d48303..d1f67818 100644
--- a/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/TransformFilterProcessor.java
+++ b/flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/TransformFilterProcessor.java
@@ -30,6 +30,7 @@ import org.slf4j.LoggerFactory;
 
 import java.lang.reflect.InvocationTargetException;
 import java.util.ArrayList;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.stream.Stream;
 
@@ -91,14 +92,12 @@ public class TransformFilterProcessor {
         List<Class<?>> argTypes = new ArrayList<>();
         String scriptExpression = transformFilter.getScriptExpression();
         List<Column> columns = tableInfo.getPreTransformedSchema().getColumns();
-        List<String> columnNames = transformFilter.getColumnNames();
+        LinkedHashSet<String> columnNames = new LinkedHashSet<>(transformFilter.getColumnNames());
         for (String columnName : columnNames) {
             for (Column column : columns) {
                 if (column.getName().equals(columnName)) {
-                    if (!argNames.contains(columnName)) {
-                        argNames.add(columnName);
-                        argTypes.add(DataTypeConverter.convertOriginalClass(column.getType()));
-                    }
+                    argNames.add(columnName);
+                    argTypes.add(DataTypeConverter.convertOriginalClass(column.getType()));
                     break;
                 }
             }
