diff --git a/src/pytorch_ie/core/document.py b/src/pytorch_ie/core/document.py
index 9be94d9..58571dc 100644
--- a/src/pytorch_ie/core/document.py
+++ b/src/pytorch_ie/core/document.py
@@ -191,6 +191,15 @@ class Annotation:
             (f.name, getattr(self, f.name)) for f in dataclasses.fields(self) if not f.compare
         )
 
+    @property
+    def comparison_fields_and_values(self) -> Tuple[Tuple[str, Any], ...]:
+        # note that we exclude the special _targets field here
+        return tuple(
+            (f.name, getattr(self, f.name))
+            for f in dataclasses.fields(self)
+            if f.compare and f.name != "_targets"
+        )
+
     @property
     def _id(self) -> int:
         # also calculate the hash over the non-comparison fields for id creation because otherwise
@@ -361,6 +370,21 @@ class Annotation:
 
         return self.copy(**overrides)
 
+    def __lt__(self, other: "Annotation") -> bool:
+        # sort only by comparison fields
+        for (name, value), (other_name, other_value) in zip(
+            self.comparison_fields_and_values, other.comparison_fields_and_values
+        ):
+            if name != other_name:
+                comparison_fields = [n for n, v in self.comparison_fields_and_values]
+                other_comparison_fields = [n for n, v in other.comparison_fields_and_values]
+                raise ValueError(
+                    f"comparison field names do not match: {comparison_fields} != {other_comparison_fields}"
+                )
+            if value != other_value:
+                return value < other_value
+        return False
+
 
 T = TypeVar("T", covariant=False, bound="Annotation")
 
