diff --git a/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/constants.py b/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/constants.py
new file mode 100644
index 0000000000..531f860e26
--- /dev/null
+++ b/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/constants.py
@@ -0,0 +1,19 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+# pylint: disable=protected-access
+
+
+class SnakeCaseFineTuningTaskTypes:
+    CHAT_COMPLETION = "chat_completion"
+    TEXT_COMPLETION = "text_completion"
+    TEXT_CLASSIFICATION = "text_classification"
+    QUESTION_ANSWERING = "question_answering"
+    TEXT_SUMMARIZATION = "text_summarization"
+    TOKEN_CLASSIFICATION = "token_classification"
+    TEXT_TRANSLATION = "text_translation"
+    IMAGE_CLASSIFICATION = "image_classification"
+    IMAGE_INSTANCE_SEGMENTATION = "image_instance_segmentation"
+    IMAGE_OBJECT_DETECTION = "image_object_detection"
+    VIDEO_MULTI_OBJECT_TRACKING = "video_multi_object_tracking"
diff --git a/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/finetuning_vertical.py b/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/finetuning_vertical.py
index f1853e0375..5d235aec54 100644
--- a/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/finetuning_vertical.py
+++ b/sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_finetuning/finetuning_vertical.py
@@ -5,11 +5,11 @@
 # pylint: disable=unused-argument
 
 from azure.ai.ml._schema._finetuning.finetuning_job import FineTuningJobSchema
+from azure.ai.ml._schema._finetuning.constants import SnakeCaseFineTuningTaskTypes
 from azure.ai.ml._schema.core.fields import NestedField, StringTransformedEnum, UnionField
 from azure.ai.ml.constants import JobType
-from azure.ai.ml._utils.utils import camel_to_snake
+from azure.ai.ml._utils.utils import snake_to_camel
 from azure.ai.ml._schema.job.input_output_entry import MLTableInputSchema, DataInputSchema, ModelInputSchema
-from azure.ai.ml._restclient.v2024_01_01_preview.models import FineTuningTaskType
 from azure.ai.ml.constants._job.finetuning import FineTuningConstants
 from azure.ai.ml._utils._experimental import experimental
 
@@ -24,19 +24,19 @@ class FineTuningVerticalSchema(FineTuningJobSchema):
     validation_data = UnionField([NestedField(MLTableInputSchema), NestedField(DataInputSchema)])
     task = StringTransformedEnum(
         allowed_values=[
-            FineTuningTaskType.CHAT_COMPLETION,
-            FineTuningTaskType.TEXT_COMPLETION,
-            FineTuningTaskType.TEXT_CLASSIFICATION,
-            FineTuningTaskType.QUESTION_ANSWERING,
-            FineTuningTaskType.TEXT_SUMMARIZATION,
-            FineTuningTaskType.TOKEN_CLASSIFICATION,
-            FineTuningTaskType.TEXT_TRANSLATION,
-            FineTuningTaskType.IMAGE_CLASSIFICATION,
-            FineTuningTaskType.IMAGE_INSTANCE_SEGMENTATION,
-            FineTuningTaskType.IMAGE_OBJECT_DETECTION,
-            FineTuningTaskType.VIDEO_MULTI_OBJECT_TRACKING,
+            SnakeCaseFineTuningTaskTypes.CHAT_COMPLETION,
+            SnakeCaseFineTuningTaskTypes.TEXT_COMPLETION,
+            SnakeCaseFineTuningTaskTypes.TEXT_CLASSIFICATION,
+            SnakeCaseFineTuningTaskTypes.QUESTION_ANSWERING,
+            SnakeCaseFineTuningTaskTypes.TEXT_SUMMARIZATION,
+            SnakeCaseFineTuningTaskTypes.TOKEN_CLASSIFICATION,
+            SnakeCaseFineTuningTaskTypes.TEXT_TRANSLATION,
+            SnakeCaseFineTuningTaskTypes.IMAGE_CLASSIFICATION,
+            SnakeCaseFineTuningTaskTypes.IMAGE_INSTANCE_SEGMENTATION,
+            SnakeCaseFineTuningTaskTypes.IMAGE_OBJECT_DETECTION,
+            SnakeCaseFineTuningTaskTypes.VIDEO_MULTI_OBJECT_TRACKING,
         ],
-        casing_transform=camel_to_snake,
+        casing_transform=snake_to_camel,
         data_key=FineTuningConstants.TaskType,
         required=True,
     )
diff --git a/sdk/ml/azure-ai-ml/azure/ai/ml/constants/_job/finetuning.py b/sdk/ml/azure-ai-ml/azure/ai/ml/constants/_job/finetuning.py
index 40a8420721..71237bcd2d 100644
--- a/sdk/ml/azure-ai-ml/azure/ai/ml/constants/_job/finetuning.py
+++ b/sdk/ml/azure-ai-ml/azure/ai/ml/constants/_job/finetuning.py
@@ -13,3 +13,17 @@ class FineTuningConstants:
     TaskType = "task"
     ModelProvider = "model_provider"
     HyperParameters = "hyperparameters"
+
+
+class FineTuningTaskTypes:
+    CHAT_COMPLETION = "ChatCompletion"
+    TEXT_COMPLETION = "TextCompletion"
+    TEXT_CLASSIFICATION = "TextClassification"
+    QUESTION_ANSWERING = "QuestionAnswering"
+    TEXT_SUMMARIZATION = "TextSummarization"
+    TOKEN_CLASSIFICATION = "TokenClassification"
+    TEXT_TRANSLATION = "TextTranslation"
+    IMAGE_CLASSIFICATION = "ImageClassification"
+    IMAGE_INSTANCE_SEGMENTATION = "ImageInstanceSegmentation"
+    IMAGE_OBJECT_DETECTION = "ImageObjectDetection"
+    VIDEO_MULTI_OBJECT_TRACKING = "VideoMultiObjectTracking"
