diff --git a/comp/core/workloadmeta/collectors/internal/containerd/containerd.go b/comp/core/workloadmeta/collectors/internal/containerd/containerd.go
index 2dd4d6b863..6200037038 100644
--- a/comp/core/workloadmeta/collectors/internal/containerd/containerd.go
+++ b/comp/core/workloadmeta/collectors/internal/containerd/containerd.go
@@ -279,13 +279,28 @@ func (c *collector) notifyInitialImageEvents(ctx context.Context, namespace stri
 		return err
 	}
 
+	mergedImages := make(map[workloadmeta.EntityID]*workloadmeta.ContainerImageMetadata)
 	for _, image := range existingImages {
-		if err := c.notifyEventForImage(ctx, namespace, image, nil); err != nil {
+		wlmImage, err := c.createOrUpdateImageMetadata(ctx, namespace, image, nil)
+		if err != nil {
 			log.Warnf("error getting information for image with name %q: %s", image.Name(), err.Error())
 			continue
 		}
+		//Image is referenced by several different names but has only one entity id (= manifest.Config.Digest).
+		//mergedImages will be refreshed by updated workloadmeta.ContainerImageMetadata
+		if wlmImage != nil {
+			mergedImages[wlmImage.EntityID] = wlmImage
+		}
+	}
+	for _, wlmImage := range mergedImages {
+		c.store.Notify([]workloadmeta.CollectorEvent{
+			{
+				Type:   workloadmeta.EventTypeSet,
+				Source: workloadmeta.SourceRuntime,
+				Entity: wlmImage,
+			},
+		})
 	}
-
 	return nil
 }
 
diff --git a/comp/core/workloadmeta/collectors/internal/containerd/image.go b/comp/core/workloadmeta/collectors/internal/containerd/image.go
index 4a0c47fb9e..17cb2b16ba 100644
--- a/comp/core/workloadmeta/collectors/internal/containerd/image.go
+++ b/comp/core/workloadmeta/collectors/internal/containerd/image.go
@@ -149,6 +149,23 @@ func (images *knownImages) getRepoDigests(imageID string) []string {
 	return res
 }
 
+// getPreferredName will return a user-friendly image name if it exists, otherwise
+// for example the name not including the digest.
+func (images *knownImages) getPreferredName(imageID string) string {
+	var res = ""
+	for ref := range images.namesByID[imageID] {
+		if res == "" && isAnImageID(ref) {
+			res = ref
+		} else if isARepoDigest(ref) {
+			res = ref // Prefer the repo digest
+			break
+		} else {
+			res = ref // Then repo tag
+		}
+	}
+	return res
+}
+
 // returns any of the existing references for the imageID. Returns empty if the
 // ID is not referenced.
 func (images *knownImages) getAReference(imageID string) string {
@@ -244,7 +261,13 @@ func (c *collector) handleImageCreateOrUpdate(ctx context.Context, namespace str
 	return c.notifyEventForImage(ctx, namespace, img, bom)
 }
 
-func (c *collector) notifyEventForImage(ctx context.Context, namespace string, img containerd.Image, sbom *workloadmeta.SBOM) error {
+// createOrUpdateImageMetadata: Create image metadata from containerd image and manifest if not already present
+// Update image metadata by adding references when existing entity is found
+// return nil when it fails to get image manifest
+func (c *collector) createOrUpdateImageMetadata(ctx context.Context,
+	namespace string,
+	img containerd.Image,
+	sbom *workloadmeta.SBOM) (*workloadmeta.ContainerImageMetadata, error) {
 	c.handleImagesMut.Lock()
 	defer c.handleImagesMut.Unlock()
 
@@ -253,7 +276,7 @@ func (c *collector) notifyEventForImage(ctx context.Context, namespace string, i
 	// Build initial workloadmeta.ContainerImageMetadata from manifest and image
 	manifest, err := images.Manifest(ctxWithNamespace, img.ContentStore(), img.Target(), img.Platform())
 	if err != nil {
-		return fmt.Errorf("error getting image manifest: %w", err)
+		return nil, fmt.Errorf("error getting image manifest: %w", err)
 	}
 
 	totalSizeBytes := manifest.Config.Size
@@ -297,6 +320,7 @@ func (c *collector) notifyEventForImage(ctx context.Context, namespace string, i
 	// of the name that includes a digest. This is just to show names that are
 	// more user-friendly (the digests are already present in other attributes
 	// like ID, and repo digest).
+	wlmImage.Name = c.knownImages.getPreferredName(wlmImage.ID)
 	existingImg, err := c.store.GetImage(wlmImage.ID)
 	if err == nil {
 		if strings.Contains(wlmImage.Name, "sha256:") && !strings.Contains(existingImg.Name, "sha256:") {
@@ -315,15 +339,21 @@ func (c *collector) notifyEventForImage(ctx context.Context, namespace string, i
 	// if the `imgMeta` object does not contain all the metadata when it is sent.
 	// We add them here to make sure they are present.
 	wlmImage.SBOM = util.UpdateSBOMRepoMetadata(wlmImage.SBOM, wlmImage.RepoTags, wlmImage.RepoDigests)
+	return &wlmImage, nil
+}
 
+func (c *collector) notifyEventForImage(ctx context.Context, namespace string, img containerd.Image, sbom *workloadmeta.SBOM) error {
+	wlmImage, err := c.createOrUpdateImageMetadata(ctx, namespace, img, sbom)
+	if err != nil {
+		return err
+	}
 	c.store.Notify([]workloadmeta.CollectorEvent{
 		{
 			Type:   workloadmeta.EventTypeSet,
 			Source: workloadmeta.SourceRuntime,
-			Entity: &wlmImage,
+			Entity: wlmImage,
 		},
 	})
-
 	return nil
 }
 
