diff --git a/tests/core/test_document.py b/tests/core/test_document.py
index d5a3c6c..ee5dbb5 100644
--- a/tests/core/test_document.py
+++ b/tests/core/test_document.py
@@ -266,6 +266,44 @@ def test_annotation_with_tuple_of_references():
     _test_annotation_reconstruction(relation, annotation_store=annotation_store)
 
 
+def test_annotation_sort():
+    @dataclasses.dataclass(eq=True, frozen=True)
+    class Dummy(Annotation):
+        a: int
+        b: Optional[int] = None
+        c: int = 0
+
+    dummy1 = Dummy(a=1, c=2)
+    dummy2 = Dummy(a=1, c=3)
+    dummy3 = Dummy(a=2, c=1)
+    dummy4 = Dummy(a=2, c=2)
+
+    assert sorted([dummy1, dummy2, dummy3, dummy4]) == [dummy1, dummy2, dummy3, dummy4]
+
+    @dataclasses.dataclass(eq=True, frozen=True)
+    class DummyWithNestedAnnotation(Annotation):
+        a: int
+        n: Dummy
+
+    dummy_nested1 = DummyWithNestedAnnotation(a=1, n=Dummy(a=1, c=2))
+    dummy_nested2 = DummyWithNestedAnnotation(a=2, n=Dummy(a=2, c=3))
+    dummy_nested3 = DummyWithNestedAnnotation(a=2, n=Dummy(a=1, c=4))
+    dummy_nested4 = DummyWithNestedAnnotation(a=1, n=Dummy(a=2, c=2))
+
+    assert sorted([dummy_nested1, dummy_nested2, dummy_nested3, dummy_nested4]) == [
+        dummy_nested1,
+        dummy_nested4,
+        dummy_nested3,
+        dummy_nested2,
+    ]
+
+    with pytest.raises(ValueError) as excinfo:
+        sorted([dummy1, dummy_nested1])
+    assert (
+        str(excinfo.value) == "comparison field names do not match: ['a', 'n'] != ['a', 'b', 'c']"
+    )
+
+
 def test_annotation_is_attached():
     @dataclasses.dataclass
     class MyDocument(Document):
