diff --git a/.gitignore b/.gitignore
index c72ed63..1e1af79 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,5 @@
 .idea
 bin
 obj
-**/*.user
\ No newline at end of file
+**/*.user
+FracturedJsonCli/scratch
diff --git a/FracturedJson/Formatter.cs b/FracturedJson/Formatter.cs
index cdbf0bd..498f258 100644
--- a/FracturedJson/Formatter.cs
+++ b/FracturedJson/Formatter.cs
@@ -156,8 +156,15 @@ public class Formatter
         foreach(var child in item.Children)
             ComputeItemLengths(child);
 
+        item.ValueLength = item.Type switch
+        {
+            JsonItemType.Null => _pads.LiteralNullLen,
+            JsonItemType.True => _pads.LiteralTrueLen,
+            JsonItemType.False => _pads.LiteralFalseLen,
+            _ => StringLengthFunc(item.Value)
+        };
+
         item.NameLength = StringLengthFunc(item.Name);
-        item.ValueLength = StringLengthFunc(item.Value);
         item.PrefixCommentLength = StringLengthFunc(item.PrefixComment);
         item.MiddleCommentLength = StringLengthFunc(item.MiddleComment);
         item.PostfixCommentLength = StringLengthFunc(item.PostfixComment);
@@ -280,7 +287,7 @@ public class Formatter
             return false;
 
         // If all items are alike, we'll want to format each element as if it were a table row.
-        var template = new TableTemplate(_pads, !Options.DontJustifyNumbers);
+        var template = new TableTemplate(_pads, Options.NumberListAlignment);
         template.MeasureTableRoot(item);
 
         // If we can't fit lots of them on a line, compact multiline isn't a good choice.  Table would likely
@@ -353,7 +360,7 @@ public class Formatter
 
         // Create a helper object to measure how much space we'll need.  If this item's children aren't sufficiently
         // similar, IsRowDataCompatible will be false.
-        var template = new TableTemplate(_pads, !Options.DontJustifyNumbers);
+        var template = new TableTemplate(_pads, Options.NumberListAlignment);
         template.MeasureTableRoot(item);
         if (!template.IsRowDataCompatible)
             return false;
@@ -602,9 +609,9 @@ public class Formatter
             else
                 InlineTableRawObject(template, item);
         }
-        else if (template.IsFormattableNumber && item.Type != JsonItemType.Null)
+        else if (template.IsNumberList)
         {
-            _buffer.Add(template.FormatNumber(item.Value));
+            template.FormatNumber(_buffer, item);
         }
         else
         {
diff --git a/FracturedJson/Formatting/PaddedFormattingTokens.cs b/FracturedJson/Formatting/PaddedFormattingTokens.cs
index 507e874..a11b9c4 100644
--- a/FracturedJson/Formatting/PaddedFormattingTokens.cs
+++ b/FracturedJson/Formatting/PaddedFormattingTokens.cs
@@ -13,6 +13,9 @@ internal class PaddedFormattingTokens
     public int CommaLen { get; }
     public int ColonLen { get; }
     public int CommentLen { get; }
+    public int LiteralNullLen { get; }
+    public int LiteralTrueLen { get; }
+    public int LiteralFalseLen { get; }
     public int PrefixStringLen { get; }
     public string DummyComma => Spaces(CommaLen);
     
@@ -63,6 +66,9 @@ internal class PaddedFormattingTokens
         CommaLen = strLenFunc(Comma);
         ColonLen = strLenFunc(Colon);
         CommentLen = strLenFunc(Comment);
+        LiteralNullLen = strLenFunc("null");
+        LiteralTrueLen = strLenFunc("true");
+        LiteralFalseLen = strLenFunc("false");
         PrefixStringLen = strLenFunc(opts.PrefixString);
     }
 
diff --git a/FracturedJson/Formatting/TableTemplate.cs b/FracturedJson/Formatting/TableTemplate.cs
index 3f51cfc..c5b6d7f 100644
--- a/FracturedJson/Formatting/TableTemplate.cs
+++ b/FracturedJson/Formatting/TableTemplate.cs
@@ -7,8 +7,9 @@ using System.Text.RegularExpressions;
 namespace FracturedJson.Formatting;
 
 /// <summary>
-/// Collects spacing information about the columns of a potential table.  Each TableTemplate corresponds do
-/// a part of a row, and they're nested recursively to match the JSON structure.
+/// Collects spacing information about the columns of a potential table.  Each TableTemplate corresponds to
+/// a part of a row, and they're nested recursively to match the JSON structure.  (Also used in complex multiline
+/// arrays to try to fit them all nicely together.)
 /// </summary>
 /// <remarks>
 /// <para> Say you have an object/array where each item would make a nice row all by itself.  We want to try to line up 
@@ -22,6 +23,11 @@ internal class TableTemplate
     /// The property name in the table that this segment matches up with. 
     /// </summary>
     public string? LocationInParent { get; private set; }
+
+    /// <summary>
+    /// The type of values in the column, if they're uniform.  There's some wiggle-room here: for instance,
+    /// true and false have different JsonItemTypes but are considered the same type for table purposes.
+    /// </summary>
     public JsonItemType Type { get; private set; } = JsonItemType.Null;
     
     /// <summary>
@@ -37,8 +43,29 @@ internal class TableTemplate
     public int MiddleCommentLength { get; private set; }
     public int PostfixCommentLength { get; private set; }
     public BracketPaddingType PadType { get; private set; } = BracketPaddingType.Simple;
-    public bool IsFormattableNumber { get; private set; }
+
+    /// <summary>
+    /// True if this is a number column and we're allowed by settings to normalize numbers (rewrite them with the same
+    /// precision), and if none of the numbers have too many digits or require scientific notation.
+    /// </summary>
+    public bool AllowNumberNormalization { get; private set; }
+
+    /// <summary>
+    /// True if this column contains only numbers and nulls.  Number columns are formatted specially, depending on
+    /// settings.
+    /// </summary>
+    public bool IsNumberList { get; private set; }
+
+    /// <summary>
+    /// Length of the value for this template when things are complicated.  For arrays and objects, it's the sum of
+    /// all the child templates' lengths, plus brackets and commas and such.  For number lists, it's the space
+    /// required to align them as appropriate.
+    /// </summary>
     public int CompositeValueLength { get; private set; }
+
+    /// <summary>
+    /// Length of the entire template, including space for the value, property name, and all comments.
+    /// </summary>
     public int TotalLength { get; private set; }
     
     /// <summary>
@@ -47,11 +74,12 @@ internal class TableTemplate
     /// </summary>
     public IList<TableTemplate> Children { get; set; } = new List<TableTemplate>();
 
-    public TableTemplate(PaddedFormattingTokens pads, bool allowReformattingNumbers)
+    public TableTemplate(PaddedFormattingTokens pads, NumberListAlignment numberListAlignment)
     {
         _pads = pads;
-        _allowReformattingNumbers = allowReformattingNumbers;
-        IsFormattableNumber = allowReformattingNumbers;
+        _numberListAlignment = numberListAlignment;
+        AllowNumberNormalization = (numberListAlignment == NumberListAlignment.Normalize);
+        IsNumberList = true;
     }
 
     /// <summary>
@@ -88,31 +116,78 @@ internal class TableTemplate
         return false;
     }
 
-    public string FormatNumber(string originalValueString)
+    /// <summary>
+    /// Added the number, properly aligned and possibly reformatted, according to our measurements.
+    /// This assumes that the segment is a number list, and therefore that the item is a number or null.
+    /// </summary>
+    public void FormatNumber(IBuffer buffer, JsonItem item)
     {
-        if (!IsFormattableNumber)
-            throw new FracturedJsonException("Logic error - attempting to format inappropriate thing as number");
+        var formatType = (_numberListAlignment is NumberListAlignment.Normalize && !AllowNumberNormalization)
+            ? NumberListAlignment.Left
+            : _numberListAlignment;
+
+        // The easy cases.  Use the value exactly as it was in the source doc.
+        switch (formatType)
+        {
+            case NumberListAlignment.Left:
+                buffer.Add(item.Value, _pads.Spaces(SimpleValueLength - item.ValueLength));
+                return;
+            case NumberListAlignment.Right:
+                buffer.Add(_pads.Spaces(SimpleValueLength - item.ValueLength), item.Value);
+                return;
+        }
+
+        // Normalize case - rewrite the number with the appropriate precision.
+        if (formatType is NumberListAlignment.Normalize)
+        {
+            if (item.Type is JsonItemType.Null)
+            {
+                buffer.Add(_pads.Spaces(_maxDigBeforeDecNorm - item.ValueLength), item.Value,
+                    _pads.Spaces(CompositeValueLength - _maxDigBeforeDecNorm));
+                return;
+            }
 
-        // Create a .NET format string, if we don't already have one.
-        if (_numberFormat == null)
+            // Create a .NET format string, if we don't already have one.
+            _numberFormat ??= $"{{0,{CompositeValueLength}:F{_maxDigAfterDecNorm}}}";
+
+            var reformattedStr = string.Format(CultureInfo.InvariantCulture, _numberFormat, double.Parse(item.Value));
+            buffer.Add(reformattedStr);
+            return;
+        }
+
+        // Decimal case - line up the decimals (or E's) but leave the value exactly as it was in the source.
+        if (item.Type is JsonItemType.Null)
         {
-            var totalWidth = _maxDigitsBeforeDecimal + _maxDigitsAfterDecimal + ((_maxDigitsAfterDecimal > 0) ? 1 : 0);
+            buffer.Add(_pads.Spaces(_maxDigBeforeDecRaw - item.ValueLength), item.Value,
+                _pads.Spaces(CompositeValueLength - _maxDigBeforeDecRaw));
+            return;
+        }
 
-            // Leave room for the literal null if we saw one in the row data.
-            if (_dataContainsNull && totalWidth < 4)
-                totalWidth = 4;
-            _numberFormat = $"{{0,{totalWidth}:F{_maxDigitsAfterDecimal}}}";
+        int leftPad;
+        int rightPad;
+        var indexOfDot = item.Value.IndexOfAny(_dotOrE);
+        if (indexOfDot > 0)
+        {
+            leftPad = _maxDigBeforeDecRaw - indexOfDot;
+            rightPad = CompositeValueLength - leftPad - item.ValueLength;
+        }
+        else
+        {
+            leftPad = _maxDigBeforeDecRaw - item.ValueLength;
+            rightPad = CompositeValueLength - _maxDigBeforeDecRaw;
         }
 
-        return string.Format(CultureInfo.InvariantCulture, _numberFormat, double.Parse(originalValueString));
+        buffer.Add(_pads.Spaces(leftPad), item.Value, _pads.Spaces(rightPad));
     }
 
+    private static readonly char[] _dotOrE = new[] { '.', 'e', 'E' };
     private readonly PaddedFormattingTokens _pads;
-    private readonly bool _allowReformattingNumbers;
-    private int _maxDigitsBeforeDecimal = 0;
-    private int _maxDigitsAfterDecimal = 0;
+    private readonly NumberListAlignment _numberListAlignment;
+    private int _maxDigBeforeDecRaw = 0;
+    private int _maxDigAfterDecRaw = 0;
+    private int _maxDigBeforeDecNorm = 0;
+    private int _maxDigAfterDecNorm = 0;
     private string? _numberFormat;
-    private bool _dataContainsNull = false;
 
     // Regex to help us distinguish between numbers that truly have a zero value - which can take many forms like
     // 0, 0.000, and 0.0e75 - and numbers too small for a 64bit float, such as 1e-500.
@@ -135,7 +210,7 @@ internal class TableTemplate
         {
             IsRowDataCompatible &= (Type is JsonItemType.True or JsonItemType.Null);
             Type = JsonItemType.True;
-            IsFormattableNumber = false;
+            IsNumberList = false;
         }
         else if (rowSegment.Type is JsonItemType.Number)
         {
@@ -144,14 +219,15 @@ internal class TableTemplate
         }
         else if (rowSegment.Type is JsonItemType.Null)
         {
-            _dataContainsNull = true;
+            _maxDigBeforeDecNorm = Math.Max(_maxDigBeforeDecNorm, _pads.LiteralNullLen);
+            _maxDigBeforeDecRaw = Math.Max(_maxDigBeforeDecRaw, _pads.LiteralNullLen);
         }
         else
         {
             IsRowDataCompatible &= (Type == rowSegment.Type || Type == JsonItemType.Null);
             if (Type is JsonItemType.Null)
                 Type = rowSegment.Type;
-            IsFormattableNumber = false;
+            IsNumberList = false;
         }
 
         // If multiple lines are necessary for a row (probably due to pesky comments), we can't make a table.
@@ -168,6 +244,7 @@ internal class TableTemplate
         if (rowSegment.Complexity >= 2)
             PadType = BracketPaddingType.Complex;
 
+        // Everything after this is moot if the column doesn't have a uniform type.
         if (!IsRowDataCompatible)
             return;
         
@@ -178,7 +255,7 @@ internal class TableTemplate
             for (var i = 0; i < rowSegment.Children.Count; ++i)
             {
                 if (Children.Count <= i)
-                    Children.Add(new(_pads, _allowReformattingNumbers));
+                    Children.Add(new(_pads, _numberListAlignment));
                 var subTemplate = Children[i];
                 subTemplate.MeasureRowSegment(rowSegment.Children[i]);
             }
@@ -203,17 +280,17 @@ internal class TableTemplate
                 var subTemplate = Children.FirstOrDefault(tt => tt.LocationInParent == rowSegChild.Name);
                 if (subTemplate == null)
                 {
-                    subTemplate = new(_pads, _allowReformattingNumbers) { LocationInParent = rowSegChild.Name };
+                    subTemplate = new(_pads, _numberListAlignment) { LocationInParent = rowSegChild.Name };
                     Children.Add(subTemplate);
                 }
                 subTemplate.MeasureRowSegment(rowSegChild);
             }
         }
-        else if (rowSegment.Type == JsonItemType.Number && IsFormattableNumber)
+        else if (rowSegment.Type == JsonItemType.Number && IsNumberList)
         {
-            // So far, everything in this column is a number that we can safely reformat.  Check to see if that's
-            // still true with this new rowSegment, and if so, measure how much space we need before and after
-            // the decimal point.
+            // So far, everything in this column is a number (or null).  We need to reevaluate whether we're allowed
+            // to normalize the numbers - write them all with the same number of digits after the decimal point.
+            // We also need to take some measurements for both contingencies.
             const int maxChars = 15;
             var parsedVal = double.Parse(rowSegment.Value);
             var normalizedStr = parsedVal.ToString("G", CultureInfo.InvariantCulture);
@@ -222,18 +299,30 @@ internal class TableTemplate
             // 1e-500 becomes 0.  In either of those cases, we shouldn't try to reformat the column.  Likewise,
             // if there are too many digits or it needs to be expressed in scientific notation, we're better off
             // not even trying.
-            IsFormattableNumber = !double.IsNaN(parsedVal)
+            AllowNumberNormalization &= !double.IsNaN(parsedVal)
                                   && !double.IsInfinity(parsedVal)
                                   && normalizedStr.Length <= maxChars
                                   && !normalizedStr.Contains('E')
                                   && (parsedVal!=0.0 || _trulyZeroValString.IsMatch(rowSegment.Value));
 
-            var indexOfDot = normalizedStr.IndexOf('.');
-            _maxDigitsBeforeDecimal =
-                Math.Max(_maxDigitsBeforeDecimal, (indexOfDot >= 0) ? indexOfDot : normalizedStr.Length);
-            _maxDigitsAfterDecimal =
-                Math.Max(_maxDigitsAfterDecimal, (indexOfDot >= 0) ? normalizedStr.Length - indexOfDot - 1 : 0);
+            // Measure the number of digits before and after the decimal point if we write it as a standard,
+            // non-scientific notation number.
+            var indexOfDotNorm = normalizedStr.IndexOf('.');
+            _maxDigBeforeDecNorm =
+                Math.Max(_maxDigBeforeDecNorm, (indexOfDotNorm >= 0) ? indexOfDotNorm : normalizedStr.Length);
+            _maxDigAfterDecNorm =
+                Math.Max(_maxDigAfterDecNorm, (indexOfDotNorm >= 0) ? normalizedStr.Length - indexOfDotNorm - 1 : 0);
+
+            // Measure the number of digits before and after the decimal point (or E scientific notation with not
+            // decimal point), using the number exactly as it was in the input document.
+            var indexOfDotRaw = rowSegment.Value.IndexOfAny(_dotOrE);
+            _maxDigBeforeDecRaw =
+                Math.Max(_maxDigBeforeDecRaw, (indexOfDotRaw >= 0) ? indexOfDotRaw : rowSegment.ValueLength);
+            _maxDigAfterDecRaw =
+                Math.Max(_maxDigAfterDecRaw, (indexOfDotRaw >= 0) ? rowSegment.ValueLength - indexOfDotRaw - 1 : 0);
         }
+
+        AllowNumberNormalization &= IsNumberList;
     }
 
     /// <summary>
@@ -243,32 +332,26 @@ internal class TableTemplate
     /// </summary>
     private void PruneAndRecompute(int maxAllowedComplexity)
     {
-        if (maxAllowedComplexity<=0)
+        if (maxAllowedComplexity <= 0 || !IsRowDataCompatible)
             Children.Clear();
         
         foreach(var subTemplate in Children)
             subTemplate.PruneAndRecompute(maxAllowedComplexity-1);
-        
-        if (!IsRowDataCompatible)
-            Children.Clear();
 
-        CompositeValueLength = SimpleValueLength;
-        if (Children.Count>0)
+        if (IsNumberList)
+        {
+            CompositeValueLength = GetNumberFieldWidth();
+        }
+        else if (Children.Count > 0)
         {
             CompositeValueLength = Children.Sum(ch => ch.TotalLength)
                                    + Math.Max(0, _pads.CommaLen * (Children.Count - 1))
                                    + _pads.ArrStartLen(PadType)
                                    + _pads.ArrEndLen(PadType);
         }
-        else if (IsFormattableNumber)
+        else
         {
-            CompositeValueLength = _maxDigitsBeforeDecimal
-                                   + _maxDigitsAfterDecimal
-                                   + ((_maxDigitsAfterDecimal > 0) ? 1 : 0);
-
-            // Allow room for null.
-            if (_dataContainsNull && CompositeValueLength < 4)
-                CompositeValueLength = 4;
+            CompositeValueLength = SimpleValueLength;
         }
 
         TotalLength = ((PrefixCommentLength > 0) ? PrefixCommentLength + _pads.CommentLen : 0)
@@ -284,4 +367,20 @@ internal class TableTemplate
             return 0;
         return 1 + Children.Max(ch => ch.GetTemplateComplexity());
     }
+
+    private int GetNumberFieldWidth()
+    {
+        if (_numberListAlignment == NumberListAlignment.Normalize && AllowNumberNormalization)
+        {
+            var normDecLen = (_maxDigAfterDecNorm > 0) ? 1 : 0;
+            return _maxDigBeforeDecNorm + normDecLen + _maxDigAfterDecNorm;
+        }
+        else if (_numberListAlignment == NumberListAlignment.Decimal)
+        {
+            var rawDecLen = (_maxDigAfterDecRaw > 0) ? 1 : 0;
+            return _maxDigBeforeDecRaw + rawDecLen + _maxDigAfterDecRaw;
+        }
+
+        return SimpleValueLength;
+    }
 }
diff --git a/FracturedJson/FracturedJsonOptions.cs b/FracturedJson/FracturedJsonOptions.cs
index 0710b8a..505cd61 100644
--- a/FracturedJson/FracturedJsonOptions.cs
+++ b/FracturedJson/FracturedJsonOptions.cs
@@ -104,11 +104,12 @@ public record FracturedJsonOptions
     /// line comments).
     /// </summary>
     public bool OmitTrailingWhitespace { get; set; } = false;
-    
+
     /// <summary>
-    /// If true, numbers won't be right-aligned with matching precision.
+    /// Controls how lists or columns of numbers (possibly with nulls) are aligned, and whether their precision
+    /// may be normalized.
     /// </summary>
-    public bool DontJustifyNumbers { get; set; } = false;
+    public NumberListAlignment NumberListAlignment { get; set; } = NumberListAlignment.Normalize;
 
     /// <summary>
     /// Number of spaces to use per indent level.  If <see cref="UseTabToIndent"/> is true, spaces won't be used but
diff --git a/FracturedJson/NumberListAlignment.cs b/FracturedJson/NumberListAlignment.cs
new file mode 100644
index 0000000..d9e35c4
--- /dev/null
+++ b/FracturedJson/NumberListAlignment.cs
@@ -0,0 +1,32 @@
+namespace FracturedJson;
+
+/// <summary>
+/// Options for how lists or columns of numbers should be aligned, and whether the precision may be changed
+/// to be consistent.
+/// </summary>
+public enum NumberListAlignment
+{
+    /// <summary>
+    /// Left-aligns numbers, keeping each exactly as it appears in the input document.
+    /// </summary>
+    Left,
+
+    /// <summary>
+    /// Right-aligns numbers, keeping each exactly as it appears in the input document.
+    /// </summary>
+    Right,
+
+    /// <summary>
+    /// Arranges the numbers so that the decimal points line up, but keeps each value exactly as it appears in the
+    /// input document.  Numbers expressed in scientific notation are aligned according to the significand's decimal
+    /// point, if any, or the "e".
+    /// </summary>
+    Decimal,
+
+    /// <summary>
+    /// Tries to rewrite all numbers in the list in regular, non-scientific notation, all with the same number of digits
+    /// after the decimal point, all lined up.  If any of the numbers have too many digits or can't be written without
+    /// scientific notation, left-alignment is used as a fallback.
+    /// </summary>
+    Normalize,
+}
diff --git a/FracturedJsonCli/.gitignore b/FracturedJsonCli/.gitignore
deleted file mode 100644
index fb188b9..0000000
--- a/FracturedJsonCli/.gitignore
+++ /dev/null
@@ -1 +0,0 @@
-scratch
diff --git a/FracturedJsonCli/Cli.cs b/FracturedJsonCli/Cli.cs
index 1d288f2..f73a176 100644
--- a/FracturedJsonCli/Cli.cs
+++ b/FracturedJsonCli/Cli.cs
@@ -38,7 +38,17 @@ namespace FracturedJsonCli
                     { "e|expand=", "always-expand depth", (int n) => options.AlwaysExpandDepth = n },
                     { "f|file=", "input from file instead of stdin", s => fileName = s },
                     { "h|help", "show this help info and exit", _ => showHelp = true },
-                    { "j|no-justify", "don't justify parallel numbers", _ => options.DontJustifyNumbers = true },
+                    {
+                        "j|justify=", "number list justification [l,r,d,n]", s =>
+                            options.NumberListAlignment = s.ToUpper() switch
+                        {
+                            "L" or "LEFT" => NumberListAlignment.Left,
+                            "R" or "RIGHT" => NumberListAlignment.Right,
+                            "D" or "DECIMAL" => NumberListAlignment.Decimal,
+                            "N" or "NORMALIZE" => NumberListAlignment.Normalize,
+                            _ => NumberListAlignment.Left,
+                        }
+                    },
                     {
                         "l|length=",
                         "maximum total line length when inlining",
diff --git a/WebFormatter/Shared/SettingsPanel.razor b/WebFormatter/Shared/SettingsPanel.razor
index 5004c35..6a108ec 100644
--- a/WebFormatter/Shared/SettingsPanel.razor
+++ b/WebFormatter/Shared/SettingsPanel.razor
@@ -1,38 +1,58 @@
 @using FracturedJson
+@using System.Data
 @inject WebFormatterState State
 
 <h2>Settings</h2>
 
-<div class="section-label">Length and Complexity</div>
-<div><input type="number" class="number" @bind="State.Options.MaxTotalLineLength" min="0" step="10">Max Total Line Length</div>
-<div><input type="number" class="number" @bind="State.Options.MaxInlineLength" min="0" step="10">Max Inline Length</div>
+<EditForm EditContext="@_editContext">
+    <div class="section-label">Length and Complexity</div>
+    <div><input type="number" class="number" @bind="State.Options.MaxTotalLineLength" min="0" step="10">Max Total Line Length</div>
+    <div><input type="number" class="number" @bind="State.Options.MaxInlineLength" min="0" step="10">Max Inline Length</div>
 
-<div><input type="number" class="number" @bind="State.Options.MaxInlineComplexity" min="-1">Max Inline Complexity</div>
-<div><input type="number" class="number" @bind="State.Options.MaxCompactArrayComplexity" min="-1">Max Compact Array Complexity</div>
-<div><input type="number" class="number" @bind="State.Options.MaxTableRowComplexity" min="-1">Max Table Row Complexity</div>
+    <div><input type="number" class="number" @bind="State.Options.MaxInlineComplexity" min="-1">Max Inline Complexity</div>
+    <div><input type="number" class="number" @bind="State.Options.MaxCompactArrayComplexity" min="-1">Max Compact Array Complexity</div>
+    <div><input type="number" class="number" @bind="State.Options.MaxTableRowComplexity" min="-1">Max Table Row Complexity</div>
 
-<div><input type="number" class="number" @bind="State.Options.MinCompactArrayRowItems" min="0">Min Compact Array Row Items</div>
-<div><input type="number" class="number" @bind="State.Options.AlwaysExpandDepth" min="-1">Always Expand Depth</div>
+    <div><input type="number" class="number" @bind="State.Options.MinCompactArrayRowItems" min="0">Min Compact Array Row Items</div>
+    <div><input type="number" class="number" @bind="State.Options.AlwaysExpandDepth" min="-1">Always Expand Depth</div>
 
-<div class="section-label">Padding</div>
-<div><input type="checkbox" @bind="State.Options.NestedBracketPadding">Nested Bracket Padding</div>
-<div><input type="checkbox" @bind="State.Options.SimpleBracketPadding">Simple Bracket Padding</div>
-<div><input type="checkbox" @bind="State.Options.ColonPadding">Colon Padding</div>
-<div><input type="checkbox" @bind="State.Options.CommaPadding">Comma Padding</div>
-<div><input type="checkbox" @bind="State.Options.CommentPadding">Comment Padding</div>
-<div><input type="checkbox" @bind="State.Options.UseTabToIndent">Use Tab To Indent</div>
-<div><input type="number" class="number" @bind="State.Options.IndentSpaces" min="0">Indent Spaces</div>
+    <div class="section-label">Padding</div>
+    <div><input type="checkbox" @bind="State.Options.NestedBracketPadding">Nested Bracket Padding</div>
+    <div><input type="checkbox" @bind="State.Options.SimpleBracketPadding">Simple Bracket Padding</div>
+    <div><input type="checkbox" @bind="State.Options.ColonPadding">Colon Padding</div>
+    <div><input type="checkbox" @bind="State.Options.CommaPadding">Comma Padding</div>
+    <div><input type="checkbox" @bind="State.Options.CommentPadding">Comment Padding</div>
+    <div><input type="checkbox" @bind="State.Options.UseTabToIndent">Use Tab To Indent</div>
+    <div><input type="number" class="number" @bind="State.Options.IndentSpaces" min="0">Indent Spaces</div>
 
-<div class="section-label">Miscellaneous</div>
-<div><input type="checkbox" @bind="State.Options.DontJustifyNumbers">Don't Justify Numbers</div>
-<div><input type="checkbox" @bind="AllowComments">Allow Comments</div>
-<div><input type="checkbox" @bind="State.Options.PreserveBlankLines">Preserve Blank Lines</div>
-<div><input type="checkbox" @bind="State.Options.AllowTrailingCommas">Allow Trailing Commas</div>
+    <div class="section-label">Miscellaneous</div>
+    <div>
+        <InputSelect @bind-Value="@State.Options.NumberListAlignment">
+            @foreach (var alignOpt in Enum.GetValues<NumberListAlignment>())
+            {
+                <option value="@alignOpt">@alignOpt</option>
+            }
+        </InputSelect>
+        <label>Number List Alignment</label>
+    </div>
+    <div><input type="checkbox" @bind="AllowComments">Allow Comments</div>
+    <div><input type="checkbox" @bind="State.Options.PreserveBlankLines">Preserve Blank Lines</div>
+    <div><input type="checkbox" @bind="State.Options.AllowTrailingCommas">Allow Trailing Commas</div>
+
+    <button type="button" @onclick="State.SetToDefaults">Reset</button>
+    &nbsp;<a href="https://github.com/j-brooke/FracturedJson/wiki/Options" target="_blank" rel="noopener">Settings Help</a>
+</EditForm>
 
-<button type="button" @onclick="State.SetToDefaults">Reset</button>
-&nbsp;<a href="https://github.com/j-brooke/FracturedJson/wiki/Options" target="_blank" rel="noopener">Settings Help</a>
 
 @code {
+    private EditContext? _editContext;
+
+    protected override void OnInitialized()
+    {
+        _editContext = new EditContext(State.Options);
+        base.OnInitialized();
+    }
+
     private bool AllowComments
     {
         get => GetAllowComments();
