diff --git a/python/llm/dev/benchmark/harness/bigdl_llm.py b/python/llm/dev/benchmark/harness/bigdl_llm.py
index f14d680ac..b370301e8 100644
--- a/python/llm/dev/benchmark/harness/bigdl_llm.py
+++ b/python/llm/dev/benchmark/harness/bigdl_llm.py
@@ -47,8 +47,8 @@ class BigDLLM(AutoCausalLM):
             if k not in self.AutoCausalLM_ARGS:
                 self.bigdl_llm_kwargs[k] = kwargs.pop(k)
 
-        self.bigdl_llm_kwargs['use_cache'] = kwargs.get('use_cache', True)
-        self.bigdl_llm_kwargs['optimize_model'] = kwargs.get('optimize_model', True)
+        self.bigdl_llm_kwargs['use_cache'] = self.bigdl_llm_kwargs.get('use_cache', True)
+        self.bigdl_llm_kwargs['optimize_model'] = self.bigdl_llm_kwargs.get('optimize_model', True)
         AutoModelForCausalLM.from_pretrained = partial(AutoModelForCausalLM.from_pretrained, **self.bigdl_llm_kwargs)
         
         kwargs['trust_remote_code'] = kwargs.get('trust_remote_code', True)
